### PR TITLE
Burn it to the ground and rebuild from the ashes

### PR DESCRIPTION
**Pull request by PinkhamSoftware**

### CODE REVIEW

File: HomesEngland.Domain/Asset.cs [hunk 1]
```diff
@@ -1,17 +1,21 @@
 using System;
 using System.Collections.Generic;
-
+using Dapper;
 namespace HomesEngland.Domain
 {
-    public class Asset
+    [Table("Assets")]
```
Comment (craigjbass) [inline HomesEngland.Domain/Asset.cs]: I would be careful of implementing Domain Objects as first class Database citizens. This couples your Domain Model (DDD) to your Database Schema. This can be especially undesirable in scenarios where there is a stakeholder responsible for defining the database schema who is different to the stakeholder responsible for implementing the business rules.  

By creating a 1:1 mapping between the two it’s harder to untangle later

Comment (craigjbass) [reply HomesEngland.Domain/Asset.cs]: One short term fix is to extract interfaces or an abstract class from this Asset that the usecases then depend on. This at least provides a well defined plugin point that’ll help avoid too much coupling to database concerns

File: HomesEngland/UseCase/GetAsset/Models/AssetOutputModel.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        public decimal? Deposit { get; set; }
+        public decimal? AgencyEquityLoan { get; set; }
+        public decimal? DeveloperEquityLoan { get; set; }
+        public decimal? ShareOfRestrictedEquity { get; set; }
+        //Calcuation
+        public int? DifferenceFromImsExpectedCompletionToHopCompletionDate { get; set; }
+
+        public AssetOutputModel(){}
+
+        public AssetOutputModel(Asset asset)
```
Comment (craigjbass) [inline HomesEngland/UseCase/GetAsset/Models/AssetOutputModel.cs]: The output model has a dependency on the Domain. This is undesirable as it creates a transient dependency on the Domain model from your UI. You can resolve this by extracting an interface to hide this concretion

File: HomesEngland/UseCase/GetAsset/Models/GetAssetResponse.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿namespace HomesEngland.UseCase.GetAsset.Models
+{
+    public class GetAssetResponse
+    {
+        public AssetOutputModel Asset { get; set; }
```
Comment (craigjbass) [inline HomesEngland/UseCase/GetAsset/Models/GetAssetResponse.cs]: This feels like an unnecessary layer. Why can’t this be flattened?

File: HomesEngland/Gateway/SqlGateway.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Data;
+using System.Threading.Tasks;
+using Dapper;
+using HomesEngland.Domain;
+
+namespace HomesEngland.Gateway
+{
+    public class SqlGateway<T, TIndex>:IGateway<T,TIndex> where T: class, IEntity<TIndex>
```
Comment (craigjbass) [inline HomesEngland/Gateway/SqlGateway.cs]: Is there a test for this? Feels like there are edge cases handled here without considering the need.

Comment (paulmorrishill) [reply HomesEngland/Gateway/SqlGateway.cs]: Agree, also suggest implementing testing to live database instance to assert entity model compatibility with the schema.

Comment (PinkhamSoftware) [reply HomesEngland/Gateway/SqlGateway.cs]: Yep agreed

File: HomesEnglandTest/UseCase/GetAsset/GetAssetTests.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+namespace HomesEnglandTest.UseCase.GetAsset
+{
+    [TestFixture]
+    public class GetAssetTests
+    {
+        public IGetAsset _classUnderTest;
+        public Mock<IEntityReader<Asset, int>> _mockGateway;
+        public GetAssetTests()
+        {
+            _mockGateway = new Mock<IEntityReader<Asset, int>>();
```
Comment (paulmorrishill) [inline HomesEnglandTest/UseCase/GetAsset/GetAssetTests.cs]: NSubstitute is about 100 times nicer than Moq

Comment (PinkhamSoftware) [reply HomesEnglandTest/UseCase/GetAsset/GetAssetTests.cs]: Thanks for the pointer I'll check it out.

File: WebApiTest/Controller/Asset/Get/AssetControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            _mockUseCase = new Mock<IGetAsset>();
+            _classUnderTest = new AssetController(_mockUseCase.Object);
+        }
+
+        [Test]
+        public async Task GivenValidRequest_ThenReturnsGetAssetResponse()
+        {
+            //arrange
+            _mockUseCase.Setup(s => s.ExecuteAsync(It.IsAny<GetAssetRequest>())).ReturnsAsync(new GetAssetResponse());
```
Comment (paulmorrishill) [inline WebApiTest/Controller/Asset/Get/AssetControllerTests.cs]: Syntax with NSubstitue would be `useCase.ExecuteAsync(Arg.Any<GetAssetRequest>()).Returns(new GetAssetResponse());`

But as you are createing the specific request below you could just use 

`useCase.ExecuteAsync(request).Returns(new GetAssetResponse());`

NSubstitute far superior

File: HomesEngland/UseCase/GetAsset/Models/AssetOutputModel.cs (hunk 2 above)
Comment (PinkhamSoftware) [reply HomesEngland/UseCase/GetAsset/Models/AssetOutputModel.cs]: Ok, yep 👍

File: HomesEngland/UseCase/GetAsset/Models/GetAssetResponse.cs (hunk 3 above)
Comment (PinkhamSoftware) [reply HomesEngland/UseCase/GetAsset/Models/GetAssetResponse.cs]: It can be but then other responses like the search response will need to return a list and other properties. So some responses would be flattened and others not.

File: HomesEngland.Domain/Asset.cs (hunk 1 above)
Comment (PinkhamSoftware) [reply HomesEngland.Domain/Asset.cs]: Yep, I'll do that. Good call 👍

File: WebApiTest/Controller/Asset/Get/AssetControllerTests.cs (hunk 6 above)
Comment (PinkhamSoftware) [reply WebApiTest/Controller/Asset/Get/AssetControllerTests.cs]: Ok that seems quite nice and a lot less verbose 👍

Comment (PinkhamSoftware) [reply WebApiTest/Controller/Asset/Get/AssetControllerTests.cs]: Yep can confirm it's quite nice to work with thanks for the pointer 🥇

File: HomesEngland/Gateway/SqlGateway.cs (hunk 4 above)
Comment (PinkhamSoftware) [reply HomesEngland/Gateway/SqlGateway.cs]: I didn't really want this PR to be about connecting up to SQL. Everything is tested up to the Gateway. I want to tackle the gateway in another PR

**Comment by PinkhamSoftware** [comment]
Leaving out the gateway for this PR. Going to address that in a different PR.
